### PR TITLE
gluon-config-mode-mesh-vpn: display tc limits in Mbit/s

### PR DESCRIPTION
**Pull request by mweinelt**
untested as I currently lack a build env and it will be a few days before I will.

### CODE REVIEW

**Comment by kaechele** [comment]
When you test please make sure values such as

- 0,5
- 0.5
- .5
- ,5

work as especially with ADSL you still see upstream rates of less than 1 Mbit/s.
From just looking through the code I'm pretty sure 0,5 and ,5 will fail at this point.

**Comment by NeoRaider** [comment]
The datatype needs to be adjusted, so the correct validator is used. We don't support the comma as a decimal point in gluon-web, but the validator will correctly show that the entered value is unacceptable.

**Comment by rubo77** [comment]
We could automagically detect that if a user enters a value above 1000 he / she seems to mean kilobyte instead of megabyte

**Comment by mweinelt** [comment]
Switching out units is a no-no. Goes against principle of least surprise.

**Comment by mweinelt** [comment]
If someone likes to pick this up, please do. I still have no development environment as NixOS and OpenWrt don't want to play nice together.

**Comment by CodeFetch** [comment]
We should add a minimal bitrate. Less than 1 mbit/s does not make sense.

**Comment by rotanid** [comment]
for a 16/1 ADSL line, less than 1mbit/s upload limit might make sense.

**Comment by CodeFetch** [comment]
If you're the only user besides the gluon node in your private network and you and a node user try to upload a file, doesn't that mean that each one of you gets about half of the bandwidth/0.5 mbit/s? Is there really a need for doing throttling to less than 0.5 mbit/s?

**Comment by rotanid** [comment]
ok, with 100kbit/s offload most freifunk communities won't work at all, especially without segmentation :D

**Comment by Adorfer** [comment]
@rotanid we miss an indicator on the status page/respondd-data telling something like "50% of upstream packets and/or upstream bandwidth used by broadcast/background traffic". 
This would be similar to "airtime-indicator" on wifi: Telling users "sorry, but you have a link which is not very productive, consider action" (e.g. release the tc limit....)
(sorry for the derail)

**Comment by rotanid** [comment]
derailing comments will be hidden soon, answer though: as you should know, this is not simple, because the node doesn't know about the available bandwidth - and it might even vary.

**Comment by Adorfer** [comment]
I assume that the node knowth about the tc limit, because that's what we are discussing here!
off course the "knowledge" will be lost/not available by the time the router is in config mode.

in other words: i oppose to setting a hard lower limit as you propose for either direction, or saying "value must be MBit/s". **Perhaps a slider with progressive scaling could be a solution.**
a **warning** in the webui would be nice if a unreasonably low value is entered (like 100kBit/s).

But what's more important would be to have in respondd/statuspage a value/warning telling _"how much  of a link is used up by management of Batman plus Broadcasts incl. RA/ND/DHCP etc)

**Comment by rubo77** [comment]
Isn't a reasonable low band with depending on the total size of the local Mesh Cloud ?

so in a very small cloud with less than fifty nodes such a band with would still be reasonable

**Comment by Adorfer** [comment]
or in a babel network. therefor a non-linear scaled slider should help to reduce input error (but I have no idea if that feasible in the current luci)

**Comment by rotanid** [comment]
@Adorfer ok, i didn't get before that you would assume "upstream bandwidth" is the configured limit (if any limit is set)
the other suggestion i don't understand, as the config mode can't depend on the size of the mesh as it isn't active at that time.

nevertheless, someone will need to step up and continue work on this (we have 97 open issues and they will only get more if people always report/ask for things but don't fix/do 'em)

File: package/gluon-config-mode-mesh-vpn/luasrc/lib/gluon/config-mode/wizard/0300-mesh-vpn.lua [hunk 1]
```diff
@@ -46,31 +46,31 @@ return function(form, uci)
 		end
 	end
 
-	o = s:option(Value, "limit_ingress", pkg_i18n.translate("Downstream (kbit/s)"))
+	o = s:option(Value, "limit_ingress", pkg_i18n.translate("Downstream (Mbit/s)"))
 	o:depends(limit, true)
-	o.default = uci:get("simple-tc", "mesh_vpn", "limit_ingress")
+	o.default = uci:get("simple-tc", "mesh_vpn", "limit_ingress") / 1024
```
Comment (CodeFetch) [inline package/gluon-config-mode-mesh-vpn/luasrc/lib/gluon/config-mode/wizard/0300-mesh-vpn.lua]: The site.conf parameters for TC are optional. The need_number-function will return nil and the parameter will thus not be added to the uci configuration in /lib/gluon/upgrade/500-mesh-vpn in that case which will result in uci:get returning nil. A division of a nil value will result in an error.

You need to check the return value first.
`o.default = uci:get("simple-tc", "mesh_vpn", "limit_ingress") or 0
o.default = o.default / 1024`

Another option is to make a default value for TC mandatory in site.conf which makes sense as we need to implicitly set a default value here.

**Comment by mweinelt** [comment]
Sorry, I currently lack motivation to finish this.

**Comment by CodeFetch** [comment]
That's a pity. I thought the comments made clear that yours is the only way that makes sense and it could be merged as is after reviews and testing.

**Comment by NeoRaider** [comment]
Will also need to be rebased onto #1743, but I can take care of that (as well as the fix for the nil setting).

**Comment by NeoRaider** [comment]
Updated with the following changes:
- rebased onto #1743
- avoid error with nil defaults
- scale by 1000 instead of 1024, so old values that were entered in kbit/s do not end up as long decimal fractions like `2000 / 1024 = 1.953125`

**Pull request merged**